src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs

```diff
@@ -189,5 +190,15 @@ private static bool SeekBackwardsAndRead(Stream stream, byte[] buffer, out int b
                 return true;
             }
         }
+
+        private static bool GetSuppressHeaderValidation()
+        {
+            if (AppContext.TryGetSwitch("System.Compression.ZipArchiveEntry.SuppressHeaderValidation", out bool suppressValidation))
```

Comment:
Do we have a real customer asking for this switch? Or are we adding it just because of we can?

In .NET Core, we are adding quirks like this only once there is a real need for them.

Reply:
As validation was a breaking change for backwards compatibility customer asked for new API with boolean argument to suppress validation: https://github.com/dotnet/runtime/issues/1094#issuecomment-582516891. @ericstj suggested AppContext Switch instead of new API

Reply:
> #1094 (comment)

This comment has important piece of information: "As far as I can see, most zip tools do not validate"

Is this a correct statement?

If yes, it means that we have different behavior than the rest of the world by default; and you can opt-in into having the same behavior as the rest world. It sounds backwards to me. If there is a a de-facto standard for something, we should follow it. Trying to be different than the rest of the world has negative value.

Reply:
Is there any other zip library that does validation like this by default?

Reply:
> This comment has important piece of information: "As far as I can see, most zip tools do not validate"
> Is this a correct statement?

I didn't really tried that customer file with any other tool, but i have looked into structure of that file and i would assume most of the tools was detecting the issue and was not extracting it, but i could tell .Net core versions before 3.0 has been extracting it.

> Is there any other zip library that does validation like this by default?

The validation added because of this issue https://github.com/dotnet/runtime/issues/27741. Where the API was extracting a tampered zip file with much smaller uncompressed size than real one. Before adding the validation i tried that invalid zip file with several tools, some of them was throwing error for that file and didn't extract it, some was extracting it up to the uncompressed size without throwing any error, some detect the error and ask for further action. I wouldn't say i tried all existing tools but from what i have tried i would say all of them validating/detecting it except .Net Core, but handling it differently.

For .Net Core it was extracting entire file without any error, and extracted file size was much more than expected uncomressed size. After discussing with @ericstj  how we would handle it within .Net Core we decided to throw error and restrict output stream up to uncompressed size. 

I would say the issue is not the validation itself but the validation logic is not perfect for Data Descriptor section reading, the issue there:
1. it has a bug which @eerhardt found while looking into the issue
2. Zip specification for Zip64 data descriptor (4 or 8 byte) is not clear and the way we using now not working for the zip files populated by some other tools (to allow Zip64 files generated with other tool we are replacing Data descriptor read logic with WPF one)
3. Other invalid zip files supposed to be failed but has being extracted with previous .Net core versions, for that case we are adding AppContext switch for backward compatibility.

Reply:
Also want to add that AppContext switch will only suppress validation, but output stream will still be restricted by uncompressed size

Reply:
Ok, thanks for the context.

I am not convinced that we need a compat switch to allow corrupted .zip files. .NET Core is compatible between versions, not bug-for-bug compatible.

Where do these corrupted zip files come from? Is there a tool or library that is generating these corrupted .zip files?

Reply:
> Where do these corrupted zip files come from? Is there a tool or library that is generating these corrupted .zip files

For issue #2 customer reported the file produced by java API
For issue #3 i think customer said he don't know the tool as it was uploaded by his clients, but i cannot find that conversation back from the issue

We can loosen the validation more to make the file for issue #3 could pass if we need to, i will try to extract that file with other popular tools to see how they handle it

Reply:
The file for issue #3 is not a zip file, it is GDTF file which stands for General Device Type Format 

> A GDTF file consists of an XML file that contains the description and several resource files such as geometry data and gobo images. The format for 3D data is 3DS and for images PNG. These files are bundled in a ZIP archive without compression.

Even though the file format is not followed the zip spec, the tools i have tried are not counting it as invalid, so i think we should also allow such files to pass validation

Reply:
> I am not convinced that we need a compat switch to allow corrupted .zip files.
>Where do these corrupted zip files come from? Is there a tool or library that is generating these corrupted .zip files?

My understanding of the reasoning to add this switch:

As @buyaa-n mentions above, there are a set of zip files that worked before, but now with our extra validation added in 3.0, we are throwing an exception and blocking loading these files. Of course it would be great if everyone would fix their bugs, but sometimes it is just not feasible to wait for some other library to fix how they are creating zip files. Additionally, the user might not even know how the zip file was created. It could just be a file lying around. Adding this switch allows a .NET Core 3.1+ app to continue loading the file, like previous versions of .NET.

See the comments by `MikeCodesDotNET` on the original issue. For example: https://github.com/dotnet/runtime/issues/1094#issuecomment-583058916.

An additional benefit of the context switch is if there are more bugs in our validation logic, users can enable this context switch to get themselves unblocked until we can ship a fix.

Reply:
After discussing with the team we decided to remove local file header vs central directory validation we added in 3.0. But restricting output by uncompressed size logic stays. 
Therefore we don't need AppContext switch too.